SharpPcap/LibPcap/CaptureFileWriterDevice.cs

```diff
@@ -77,140 +77,53 @@ public override string Description
         {
-        }
+            if (!Opened)
+                return;
 
-        /// <summary>
-        /// Constructor
-        /// </summary>
-        /// <param name="device">
-        /// A <see cref="LibPcapLiveDevice"/>
-        /// </param>
-        /// <param name="captureFilename">
-        /// A <see cref="string"/>
-        /// </param>
-        /// <param name="mode">
-        /// A <see cref="FileMode"/>
-        /// </param>
-        public CaptureFileWriterDevice(LibPcapLiveDevice device,
-                                       string captureFilename,
-                                       FileMode mode) :
-            this((PacketDotNet.LinkLayers)LibPcapSafeNativeMethods.pcap_datalink(device.PcapHandle),
-                 LibPcapSafeNativeMethods.pcap_snapshot(device.PcapHandle),
-                 captureFilename,
-                 mode)
+            base.Close();
 
-        {
+            // close the dump handle
+            if (m_pcapDumpHandle != IntPtr.Zero)
+            {
+                LibPcapSafeNativeMethods.pcap_dump_close(m_pcapDumpHandle);
+                m_pcapDumpHandle = IntPtr.Zero;
+            }
         }
 
         /// <summary>
-        /// Constructor
+        /// Open the device
         /// </summary>
-        /// <param name="linkLayerType">
-        /// A <see cref="PacketDotNet.LinkLayers"/>
-        /// </param>
-        /// <param name="snapshotLength">
-        /// A <see cref="Nullable{T}"/> of <see cref="int"/>
-        /// </param>
-        /// <param name="captureFilename">
-        /// A <see cref="string"/>
-        /// </param>
-        /// <param name="mode">
-        /// A <see cref="FileMode"/>
-        /// </param>
-        public CaptureFileWriterDevice(PacketDotNet.LinkLayers linkLayerType,
-                                       int? snapshotLength,
-                                       string captureFilename,
-                                       FileMode mode)
+        public override void Open(DeviceConfiguration configuration)
         {
-            m_pcapFile = captureFilename;
+            m_pcapFile = configuration.Filename;
 
             // append isn't possible without some difficulty and not implemented yet
-            if (mode == FileMode.Append)
+            if (configuration.Filemode == FileMode.Append)
             {
                 throw new InvalidOperationException("FileMode.Append is not supported, please contact the developers if you are interested in helping to implementing it");
             }
 
-            if (!snapshotLength.HasValue)
-            {
-                snapshotLength = Pcap.MAX_PACKET_SIZE;
-            }
-            else if (snapshotLength > Pcap.MAX_PACKET_SIZE)
+            if (configuration.Snaplen > Pcap.MAX_PACKET_SIZE)
             {
-                throw new InvalidOperationException("snapshotLength > Pcap.MAX_PACKET_SIZE");
+                throw new InvalidOperationException("Snaplen > Pcap.MAX_PACKET_SIZE");
```

Comment:
Use the `Configure` helper method (make it protected instead of private), or use `base.Open`

Reply:
Isn't the Configure helper method in LibPcapLivePcapDevice? Should I move it down into PcapDevice for reuse?

Reply:
Yes possible

Reply:
@kayoub5 for the case where we call CreateFileWriterDevice with an invalid snaplen and we do a DeviceConfiguration.RaiseConfigurationFailed() it doesn't look like we throw an exception. How do we let the user know there is an issue with their configuration?

Reply:
Right now user needs to subscribe to `ConfigurationFailed` event.
```cs
var configuration = new DeviceConfiguration();
configuration.ConfigurationFailed += (s, e) => 
    if (e.Property == "Snaplen") {
        // the device was Opened, but setting Snaplen failed
    }
}
device.Open(configuration);
```
we could add property `DeviceConfiguration.ThrowOnFailure` that makes the `Open()` method throw, for users who really want every value of their configuration to be accepted, not just the ones that are supported.